auth-providers/oauth2-provider/src/main/java/tech/ydb/auth/OAuth2TokenExchangeProvider.java

```diff
@@ -0,0 +1,320 @@
+
+        OAuth2Rpc(ExecutorService executor) {
+            this.executor = executor;
+            this.client = HttpClients.createDefault();
+        }
+
+        @Override
+        public void close() {
+            try {
+                this.client.close();
+            } catch (IOException e) {
+                logger.error("io exception on closing of http client", e);
+            }
+        }
+
+        private Token updateToken() throws IOException {
+            if (expiredAt == null || clock.instant().isAfter(expiredAt)) {
+                tokenValue = tokenSource.getToken();
+                expiredAt = clock.instant().plusSeconds(tokenSource.getExpireInSeconds());
+            }
+            HttpPost post = new HttpPost(endpoint);
+            post.setEntity(buildHttpForm());
+            CloseableHttpResponse response = client.execute(post);
+
+            int httpCode = response.getStatusLine().getStatusCode();
+            if (httpCode != 200) {
+                StatusCode code;
+                switch (httpCode) {
+                    case 400:
+                        code = StatusCode.BAD_REQUEST;
+                        break;
+                    case 401:
+                    case 403:
+                        code = StatusCode.UNAUTHORIZED;
+                        break;
+                    case 404:
+                        code = StatusCode.NOT_FOUND;
+                        break;
+                    case 500:
+                        code = StatusCode.INTERNAL_ERROR;
+                        break;
+                    case 503:
+                    case 504:
+                        code = StatusCode.UNAVAILABLE;
+                        break;
+                    default:
+                        code = StatusCode.CLIENT_INTERNAL_ERROR;
+                        break;
+                }
+                String message = "Cannot get OAuth2 token with code " + httpCode
+                        + "[" + response.getStatusLine().getReasonPhrase() + "]";
+                throw new UnexpectedResultException(message, Status.of(code));
+            }
+
+            try (Reader reader = new InputStreamReader(response.getEntity().getContent())) {
+                OAuth2Response json = GSON.fromJson(reader, OAuth2Response.class);
+
+                if (!"Bearer".equals(json.getTokenType())) {
+                    throw new UnexpectedResultException(
+                            "OAuth2 token exchange: unsupported token type: " + json.getTokenType(),
+                            Status.of(StatusCode.INTERNAL_ERROR)
+                    );
+                }
+                if (json.getExpiredIn() == null || json.getExpiredIn() <= 0) {
+                    throw new UnexpectedResultException(
+                            "OAuth2 token exchange: incorrect expiration time: " + json.getExpiredIn(),
+                            Status.of(StatusCode.INTERNAL_ERROR)
+                    );
+                }
+
```

Comment:
It will also be a better practive to check scope parameter, because we can receive token with different scope and it will be an unexpected issue later